### DISCUSSION TITLE
Build with cmake file on Raspberry Pi 4 Failed

### DISCUSSION THREAD

**Discussion by christophezeinaty** [larq/compute-engine#783] [General]
Hello I am trying to build my code on Raspberry pi4 to test my model inference and the build is failing, I have the following error : 

`/tmp/ccoKBorx.s:7901: Error: selected processor does not support `vudot.u8 q4,q8,d4[0]' in ARM mode`

`make[2]: *** [_deps/xnnpack-build/CMakeFiles/microkernels-prod.dir/build.make:199: _deps/xnnpack-build/CMakeFiles/microkernels-prod.dir/src/amalgam/gen/neondot.c.o] Error 1`

`make[1]: *** [CMakeFiles/Makefile2:7921: _deps/xnnpack-build/CMakeFiles/microkernels-prod.dir/all] Error 2`

this code has been tested on my pc and on a Jetson Xavier AGX and it works well. Please not that I am building directly on the RPI4 and I am using a RPI4 via ssh there is no installed OS

**Comment by Tombana** [comment]
This is a known issue with older version of `gcc`. What version are you using? You can check this by running `gcc --version`.

**Reply by christophezeinaty** [reply]
I am using this gcc (Raspbian 10.2.1-6+rpi1) 10.2.1 20210110. And now I realized that the headless OS I am using is 32bits, so is it better to install a new one based on 64bits to benefit from LCE ops ?

**Reply by Tombana** [reply]
LCE will work on 32-bits but indeed it has been more optimized for a 64-bit system.

EDIT:
gcc 10.2 is known to have this bug. The bug was fixed in fixed in 2022 and even backported to gcc 9, but the gcc on Raspbian is not updated with the patch. (Bug is due to an incompatibility between the binutils assembler and gcc, see [this issue](https://github.com/google/XNNPACK/issues/1465))

**Reply by christophezeinaty** [reply]
ok thanks but is it normal that I have this issue with the version my gcc 10.2.1 20210110 ?

**Reply by Tombana** [reply]
Yes. (I was editing my post above, should've made a new post)

**Reply by christophezeinaty** [reply]
so I should install the gcc 9 ? (seems like this is the latest version 10.2 and its not working)

**Reply by Tombana** [reply]
I think the easiest is to install a new 64-bit OS, I think that compiler will work. I do have a possible fix for the 32-bit OS that I applied on a Raspberry Pi in the past, but it is a bit hacky, so use it at your own risk:

Downloaded `binutils <= 2.33.1` from [this website](https://snapshot.debian.org/package/binutils/2.33-1/). On that page you have to download `binutils-arm-linux-gnueabi_2.33-1_armel.deb` as well as `libbinutils_2.33-1_armhf.deb`. Do **not** install them with `apt` but simply unzip them somewhere. Then backup the system's assembler `/usr/bin/arm-linux-gnueabihf-as` and overwrite it with a symlink to the unzipped `arm-linux-gnueabihf-as`. To make it find the libraries from `libbinutils`, add this to `.bashrc`:
```bash
export LD_LIBRARY_PATH="<path_to_libbinutils>/usr/lib/arm-linux-gnueabihf"
```

**Reply by christophezeinaty** [reply]
I ended up doing the easy solution, I installed the new 64bit OS and it worked thank you !
